### PR TITLE
Viewport interaction

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/openloco/industry.cpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ namespace openloco
                             if (bl == 0 || bl != obj->var_EA)
                             {
                                 var_DB++;
-                                if ((!(obj->var_E4 & 0x10000000) && (surface->data()[4] & 0xE0) == 0) || find_5(surface))
+                                if ((!(obj->flags & 0x10000000) && (surface->data()[4] & 0xE0) == 0) || find_5(surface))
```
Comment (marijnvdwerf) [inline src/openloco/industry.cpp]: Add a constant for this flag 0x10000000. Are we sure about this value? [locotool](https://github.com/jawsper/locotool/blob/master/libloco/objects.py#L800) only goes up to `800000`

File: src/openloco/map/tile.cpp [hunk 2]
```diff
@@ -27,7 +27,7 @@ building_object* building_element::object() const
     return objectmgr::get<building_object>(object_id());
 }
 
-industry* industry_element::industry() const
+industry* tree_element::industry() const
```
Comment (marijnvdwerf) [inline src/openloco/map/tile.cpp]: obviously wrong

File: src/openloco/things/vehicle.cpp [hunk 3]
```diff
@@ -1093,26 +1093,26 @@ void openloco::vehicle::steam_puffs_animation_update(uint8_t num, int8_t var_05)
 
         for (auto& el : tile)
         {
-            if (itemFound && !(el.flags() & ((1 << 5) | (1 << 4))))
+            if (itemFound && !(el.flags() & (map::element_flags::flag_5 | map::element_flags::flag_4)))
```
Comment (marijnvdwerf) [inline src/openloco/things/vehicle.cpp]: Just use `is_flag_5` and `is_flag_4`

File: src/openloco/window.cpp [hunk 4]
```diff
@@ -387,16 +387,19 @@ namespace openloco::ui
 
     // 0x00459E54
     // TODO: needs expansion in terms of (output) parameters.
-    static void get_map_coordinates_from_pos(int32_t screenX, int32_t screenY, int32_t flags, int16_t* x, int16_t* y)
+    viewport_interaction::InteractionItem get_map_coordinates_from_pos(int32_t screenX, int32_t screenY, int32_t flags, int16_t* x, int16_t* y, void** dest)
     {
         registers regs;
         regs.ax = screenX;
-        regs.cx = screenY;
+        regs.bx = screenY;
```
Comment (marijnvdwerf) [inline src/openloco/window.cpp]: @AaronVanGeffen Does this change fix your 'zoom to cursor' woes?

Comment (AaronVanGeffen) [reply src/openloco/window.cpp]: Yes!! Just tested this and I'm baffled — this is great news! Thank you so much.

Shall we tackle it in a different PR? We could enable it by default and all that, too.

File: src/openloco/window.cpp [hunk 5]
```diff
@@ -387,16 +387,19 @@ namespace openloco::ui
...
         regs.edx = flags;
         call(0x00459E54, regs);
 
         *x = regs.ax;
         *y = regs.cx;
```
Comment (AaronVanGeffen) [inline src/openloco/window.cpp]: Does this line need changing too, or is this perhaps where my original confusion (b2e1f298) stems from?

File: src/openloco/window.cpp (hunk 4 above)
Comment (marijnvdwerf) [reply src/openloco/window.cpp]: Yeah, this PR contains lots of stuf that would probably be better extracted

File: src/openloco/window.cpp (hunk 5 above)
Comment (marijnvdwerf) [reply src/openloco/window.cpp]: I think Sawyer uses different default registers for different coordinate systems. The 3D coordinates tend to go in `ax`, `cx`, and occasionally `dx` for height, whereas the 2D screen-space coordinates tend to go in `ax` and `bx`. That said, I haven't looked into this function enough to give a solid answer.

Comment (AaronVanGeffen) [reply src/openloco/window.cpp]: Good to know! Thanks, I'll keep this in mind.

**Comment by duncanspumpkin** [comment]
I've restarted work on this in https://github.com/duncanspumpkin/OpenLoco/tree/viewport_interact. I'm tempted to split this into two PR's one for viewport_interaction.cpp and the other for mouse_input.cpp. Both are long and complex.

**Comment by AaronVanGeffen** [comment]
Superseded by #546 and #547.

**Pull request closed without merging**